### PR TITLE
[nnc] Fix float->bool conversion on cpu

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57798 [nnc] Fix float->bool conversion on cpu**

Our instruction sequence was just plain wrong, instead of `fcmp une %x, +0.0`
(unordered equal 0.0) we were doing `fcmp uno`, which is just an unordered check
(i.e., is either side NaN).

Differential Revision: [D28276269](https://our.internmc.facebook.com/intern/diff/D28276269/)

### CODE REVIEW

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 1]
```diff
@@ -1087,7 +1087,7 @@ class TensorExprFuser {
 
     // A hook to optimizations limitter to allow bisecting the pass
     if (!JIT_OPT_ALLOWED) {
-      return false;
+      REQ(false && "jit opt limited");
```
Comment (navahgar) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:1090]: Does this belong to this PR?

**Review by navahgar — APPROVED** [review]
Good catch. LGTM

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 1 above)
Comment (bertmaher) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:1090]: Yeah it's not strictly related but it was useful to me while bisecting so I thought I might as well leave it.

Comment (ZolotukhinM) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:1090]: Nit: probably could've rewritten it as `REQ(JIT_OPT_ALLOWED && "jit opt limited")`.

Comment (bertmaher) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:1090]: Yeah good point that's cleaner.  I'll change to that!

Comment (ZolotukhinM) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:1090]: I'm actually not sure the macros magic is composable, but would be cool if it works :)

**Review by eellison — COMMENTED** [review]
Oop, my bad! Do you know why this wasnt caught by the test? https://github.com/pytorch/pytorch/pull/48976/files#diff-5b331823d79fc13c5436ad744a4856c5ef0acd9403ca3fa9d4a12f8ec69cdc8aR1281 I have float->bool tests there..

**Comment by bertmaher** [comment]
> Oop, my bad! Do you know why this wasnt caught by the test? https://github.com/pytorch/pytorch/pull/48976/files#diff-5b331823d79fc13c5436ad744a4856c5ef0acd9403ca3fa9d4a12f8ec69cdc8aR1281 I have float->bool tests there..

Dang, this is a weakness in my `data_for` function that we might want to systematically remedy.  For `float` dtypes it generates data with `rand`, but this conversion only fails with a precise value of 0.0, which is unlikely as a result of `rand`.

**Pull request closed without merging**